src/spartan_router.erl

```diff
@@ -75,7 +76,33 @@ find_upstream(_Name, [<<"zk">>|_]) ->
     erldns_resolvers();
 find_upstream(_Name, [<<"spartan">>|_]) ->
     erldns_resolvers();
-find_upstream(Name, _Labels) ->
+find_upstream(Name, Labels) ->
+    case find_custom_upstream(Labels) of
+        [] ->
+            find_default_upstream(Name);
+        Resolvers ->
+            lager:debug("resolving ~p with custom upstream: ~p", [Labels, Resolvers]),
+            Resolvers
+    end.
+
+-spec(find_custom_upstream(Labels :: [binary()]) -> [{string(), inet:port_number()}]).
+find_custom_upstream(QueryLabels) ->
+    ForwardZones = spartan_config:forward_zones(),
+
+    UpstreamFilter =
```

Comment:
Move this to its own function for readability. Somthing like mk_fun_upstream_filter..